### DISCUSSION TITLE
`breakdown` output improvements?

### DISCUSSION THREAD

**Discussion by aliscott** [infracost/infracost#1071] [Ideas & Feature Requests]
I've been collating feedback on how we improve our default breakdown table output and thought I'd start a discussion around this. I'm posting ideas here as comments. Feel free to add any more ideas and upvote/reply to existing ones.

**Comment by aliscott** [comment]
### Show the total cost of each resource

Add the total cost of each resource, so users can grep for a resource name. Example below:

```
 Name                                                   Monthly Qty  Unit         Monthly Cost

 aws_instance.web_app                                                                  $742.64
 ├─ Instance usage (Linux/UNIX, on-demand, m5.4xlarge)          730  hours             $560.64
 ├─ root_block_device
 │  └─ Storage (general purpose SSD, gp2)                        50  GB                  $5.00
 └─ ebs_block_device[0]
    ├─ Storage (provisioned IOPS SSD, io1)                    1,000  GB                $125.00
    └─ Provisioned IOPS                                         800  IOPS               $52.00

 aws_lambda_function.hello_world                                                       $436.67
 ├─ Requests                                                    100  1M requests        $20.00
 └─ Duration                                             25,000,000  GB-seconds        $416.67

 OVERALL TOTAL                                                                       $1,179.31
```

Does this make the output confusing since now everything in the `Monthly Cost` column. doesn't sum to the total cost?

**Reply by jgrumboe** [reply]
Instead of having this kind of sub-totals I would go for the "Show costs at the resource or module level".

**Reply by dekimsey** [reply]
Just spit-balling, perhaps move totals out a column.

```
 Name                                                   Monthly Qty  Unit         Monthly Cost         Totals

 aws_instance.web_app                                                                                    $742.64
 ├─ Instance usage (Linux/UNIX, on-demand, m5.4xlarge)          730  hours             $560.64
 ├─ root_block_device
 │  └─ Storage (general purpose SSD, gp2)                        50  GB                  $5.00
 └─ ebs_block_device[0]
    ├─ Storage (provisioned IOPS SSD, io1)                    1,000  GB                $125.00
    └─ Provisioned IOPS                                         800  IOPS               $52.00

 aws_lambda_function.hello_world                                                                         $436.67
 ├─ Requests                                                    100  1M requests        $20.00
 └─ Duration                                             25,000,000  GB-seconds        $416.67

 OVERALL TOTAL                                                                                         $1,179.31

```

**Comment by aliscott** [comment]
### Show costs at the resource or module level

Some previous discussion: [https://github.com/infracost/infracost/issues/322](https://github.com/infracost/infracost/issues/322)

Currently we show the full breakdown of costs for each price point for a resource. For some resources, like S3 this can include many different price points. We could add an option for showing the costs at the resource or terraform module level, so instead of showing the full breakdown we just show the total cost for that resource or terraform module

Resource example:

```
 Name                             Monthly Cost

 aws_instance.web_app                  $742.64
 aws_lambda_function.hello_world       $436.67

 OVERALL TOTAL                       $1,179.31
```

Terraform module example:

```
 Name                             Monthly Cost

 module.vpc                            $167.81
 module.base                           $233.43
 module.ec2                            $967.67

 OVERALL TOTAL                       $1,368.91
```

If we do this:

1. What should the flag name and values be? Initial flag name ideas: `--level`, `--breakdown-level`, `--aggregate-by`. Value ideas: `components, resources, terraform_modules`
2. What should the default breakdown level be? Individual cost components (current), resources, terraform modules or something else?

**Reply by jgrumboe** [reply]
That's a nice idea!
ad 1. Since the command is already `breakdown` I wouldn't choose `--breakdown-level` because of repetition. `--level` is short and would be fine for me.
ad 2. I would keep the current cost component level as a default.

**Comment by aliscott** [comment]
### Hide small costs by default

By default combine any costs that are <$1 into a "Other costs" cost component. These can be expanded using a flag (e.g. `--verbose` or `--full`)

**Reply by PatMyron** [reply]
>combine any costs that are <$1

Could additionally consider percentage-based, probably makes sense for the threshold to scale a bit with the overall stack cost

---

Also would be nice if dollar/percent threshold was configurable

**Comment by aliscott** [comment]
### Sort the costs by most expensive first

Sort the resources by most expensive first and sort the sub-resources and cost components inside each resource by most expensive first.
